File: src/formatters/render-pango.ts
import { ONE_DARK } from '../theme';
import type { ColorToken, Line, Segment } from './segments';

/** Theme color token → Pango hex. Shared so every Pango caller escapes identically. */
export const HEX_BY_TOKEN: Record<ColorToken, string> = {
  green: ONE_DARK.green,
  yellow: ONE_DARK.yellow,
  orange: ONE_DARK.orange,
  red: ONE_DARK.red,
  comment: ONE_DARK.comment,
  text: ONE_DARK.text,
  textBright: ONE_DARK.textBright,
  muted: ONE_DARK.muted,
  magenta: ONE_DARK.magenta,
  cyan: ONE_DARK.cyan,
  blue: ONE_DARK.blue,
  brightBlue: ONE_DARK.brightBlue,
};

function escapeXml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/'/g, '&#39;')
    .replace(/"/g, '&quot;');
}

/**
 * Wrap text in a Pango `<span>`, XML-escaping the text. This is the single escape
 * boundary for Pango output — all bar text and tooltips must route through here
 * (or `renderPango`) so untrusted provider data can never inject markup.
 */
export function span(color: string, text: string, bold = false): string {
  return `<span foreground='${color}'${bold ? " weight='bold'" : ''}>${escapeXml(text)}</span>`;
}

function renderSegment(seg: Segment): string {
  if (seg.raw) return seg.text;
  return span(HEX_BY_TOKEN[seg.color], seg.text, seg.bold ?? false);
}

function renderLine(line: Line): string {
  return line.map(renderSegment).join('');
}

/** Render a list of Lines to a multi-line Pango markup string. */
export function renderPango(lines: Line[]): string {
  return lines.map(renderLine).join('\n');
}

File: src/formatters/render_pango.rs
//! ÚNICO ponto de XML-escape do Pango. Toda string de provider que entra em markup
//! Pango passa por `span` (que escapa). Segments `raw` saem verbatim — nunca passe
//! dados não-confiáveis por `raw`.

use super::segments::{Line, Segment};

/// Escapa os 5 entities XML. ORDEM importa: `&` primeiro (senão double-escape).
pub fn escape_xml(text: &str) -> String {
    text.replace('&', "&amp;")
        .replace('<', "&lt;")
        .replace('>', "&gt;")
        .replace('\'', "&#39;")
        .replace('"', "&quot;")
}

/// `<span foreground='{hex}'[ weight='bold']>{escape(text)}</span>` — aspas simples.
pub fn span(hex: &str, text: &str, bold: bool) -> String {
    let weight = if bold { " weight='bold'" } else { "" };
    format!(
        "<span foreground='{hex}'{weight}>{}</span>",
        escape_xml(text)
    )
}

fn render_segment(seg: &Segment) -> String {
    if seg.raw {
        return seg.text.to_string();
    }
    span(seg.color.hex(), &seg.text, seg.bold)
}

fn render_line(line: &Line) -> String {
    line.iter().map(render_segment).collect()
}

/// Renderiza linhas em markup Pango multi-linha.
pub fn render_pango(lines: &[Line]) -> String {
    lines.iter().map(render_line).collect::<Vec<_>>().join("\n")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::formatters::segments::Segment;
    use crate::theme::ColorToken;

    #[test]
    fn escape_all_five_entities_in_order() {
        assert_eq!(escape_xml("a&b<c>d'e\"f"), "a&amp;b&lt;c&gt;d&#39;e&quot;f");
    }

    #[test]
    fn escape_ampersand_first_no_double_escape() {
        // se '&' não fosse primeiro, "<" viraria "&lt;" e o "&" seria re-escapado
        assert_eq!(escape_xml("<"), "&lt;");
        assert_eq!(escape_xml("&lt;"), "&amp;lt;");
    }

    #[test]
    fn span_format_single_quotes() {
        assert_eq!(
            span("#98c379", "hi", false),
            "<span foreground='#98c379'>hi</span>"
        );
    }

    #[test]
    fn span_bold_adds_weight() {
        assert_eq!(
            span("#e06c75", "x", true),
            "<span foreground='#e06c75' weight='bold'>x</span>"
        );
    }

    #[test]
    fn span_escapes_text() {
        assert_eq!(
            span("#c0c9d4", "a<b", false),
            "<span foreground='#c0c9d4'>a&lt;b</span>"
        );
    }

    #[test]
    fn render_pango_raw_bypasses_span_and_escape() {
        let line = vec![
            Segment::new("ok", ColorToken::Green),
            Segment::raw_text(" <sep> "),
        ];
        let out = render_pango(&[line]);
        assert_eq!(out, "<span foreground='#98c379'>ok</span> <sep> ");
    }

    #[test]
    fn render_pango_joins_lines_with_newline() {
        let l1 = vec![Segment::new("a", ColorToken::Text)];
        let l2 = vec![Segment::new("b", ColorToken::Text)];
        let out = render_pango(&[l1, l2]);
        assert!(out.contains('\n'));
        assert_eq!(out.lines().count(), 2);
    }
}
